### ISSUE TITLE
Make all KotlinPoetMetadataSpec extension functions public

### ISSUE THREAD

**Issue by Legion2** [square/kotlinpoet#1165]
For interop with `kotlinx.metadata.Km...` types there is currently only `KmClass.toTypeSpec` and `KmClass.toFileSpec` publicly accessible. However the other `to...Spec` extension functions are all private in https://github.com/square/kotlinpoet/blob/b4d77c02bf8c024610c4b9c04205d94e65b40dee/interop/kotlinx-metadata/src/main/kotlin/com/squareup/kotlinpoet/metadata/specs/KotlinPoetMetadataSpecs.kt. They would be very useful, because currently I have to implement a visitor which uses `Km` and `Spec` types in parallel and match them. Because only the `Km` types allow to inspect the types and the `Spec` can create a valid source representation.

**Comment by ZacSweers** [comment]
Can you be more specific about some of these use cases? We don't really intend to offer these APIs to be public and you can't just create enclosed elements without the context of the enclosing elements. The API is quite intentionally limited here.

**Comment by Legion2** [comment]
I missed that `TypeName` is a sealed class, so all information I need is available there. And I don't need the KmValueParameter and KmType anymore.

Now I was able to replace the `Km...` types with the Spec types, however one remaining problem is that `TypeSpec.primaryConstructor` is always null if I use `metadata.toKmClass().toTypeSpec(null)` to create the TypeSpec.

Here is my use case:

I want to generate a android and ios specific wrapper for a common kotlin class in a multiplatform project and therefore need to copy the functions with parameters from the original class. I use `KmClass.toTypeSpec` to create the TypeSpec of the original class and extract the functions and their parameter from the Spec and write it to the new platform specific wrapper source file. However, some parameter types have to be mapped to different platform specific types. For example, on android the `List<T>` must be mapped to `ReadableArray` type on android. For this mapping I need the type information of the function parameters, however the subclasses of `TypeName` are in a differnt source file, so I did not see them at first. Also they are not data classes and they do not override `equals` function so I thought I can not compare them to e.g. `List<T>`. Therefore i did this comparison using the `Km...` types.
